### PR TITLE
MINOR: Add read tests for RemoteLogManager

### PR DESCRIPTION
**Pull request by nikramakrishnan**
This PR adds test cases for `RemoteLogManager.read`, including cases to verify that the behaviour of remote fetch is consistent with that of local fetch when `RemoteStorageFetchInfo.minOneMessage` is set to true.

### Committer Checklist (excluded from commit message)
- [] Verify design and implementation 
- [] Verify test coverage and CI build status

### CODE REVIEW

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java [hunk 1]
```diff
@@ -1330,7 +1330,7 @@ private Optional<RemoteLogSegmentMetadata> findNextSegmentMetadata(RemoteLogSegm
                 : Optional.empty();
     }
 
-    private RecordBatch findFirstBatch(RemoteLogInputStream remoteLogInputStream, long offset) throws IOException {
+    protected RecordBatch findFirstBatch(RemoteLogInputStream remoteLogInputStream, long offset) throws IOException {
```
Comment (divijvaidya) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: default access? Also, in Kafka code base we add a `// Visible for testing` comment whenever we change access due to tests

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 2]
```diff
@@ -1361,6 +1371,188 @@ private Map<Integer, Long> truncateAndGetLeaderEpochs(List<EpochEntry> entries,
...
+        when(remoteStorageManager.fetchLogSegment(any(RemoteLogSegmentMetadata.class), anyInt()))
+                .thenAnswer(a -> fileInputStream);
+        when(mockLog.leaderEpochCache()).thenReturn(Option.apply(cache));
+
+        FetchRequest.PartitionData partitionData = new FetchRequest.PartitionData(
+                Uuid.randomUuid(), 0, 0, 0, Optional.empty()
+        );
+
+        RemoteStorageFetchInfo fetchInfo = new RemoteStorageFetchInfo(
+                0, false, tp, partitionData, FetchIsolation.TXN_COMMITTED, false
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java:1766]: please move "fetchOffset" into a var and then use the same var in assertion. It helps readibility for the test since it is easy to make out that we are expecting the same value of fetchOffset that we initially provided to read() function.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 3]
```diff
@@ -1361,6 +1371,188 @@ private Map<Integer, Long> truncateAndGetLeaderEpochs(List<EpochEntry> entries,
...
+            }
+
+            protected RecordBatch findFirstBatch(RemoteLogInputStream remoteLogInputStream, long offset) throws IOException {
+                return null;
+            }
+        }) {
+            FetchDataInfo fetchDataInfo = remoteLogManager.read(fetchInfo);
+            assertEquals(0, fetchDataInfo.fetchOffsetMetadata.messageOffset);
+            assertEquals(MemoryRecords.EMPTY, fetchDataInfo.records);
+            assertTrue(fetchDataInfo.abortedTransactions.get().isEmpty());
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java:1805]: Please first assert that `fetchDataInfo.abortedTransactions` is present otherwise we will get a NullPointerException here instead of a properly failing test.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 4]
```diff
@@ -1361,6 +1371,188 @@ private Map<Integer, Long> truncateAndGetLeaderEpochs(List<EpochEntry> entries,
...
+            protected int lookupPositionForOffset(RemoteLogSegmentMetadata remoteLogSegmentMetadata, long offset) {
+                return 1;
+            }
+
+            protected RecordBatch findFirstBatch(RemoteLogInputStream remoteLogInputStream, long offset) throws IOException {
+                return null;
+            }
+        }) {
+            FetchDataInfo fetchDataInfo = remoteLogManager.read(fetchInfo);
+            assertEquals(0, fetchDataInfo.fetchOffsetMetadata.messageOffset);
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: please assert that firstEntryIncomplete is false as well.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 5]
```diff
@@ -1361,6 +1371,188 @@ private Map<Integer, Long> truncateAndGetLeaderEpochs(List<EpochEntry> entries,
         return myCheckpoint.read().stream().collect(Collectors.toMap(e -> e.epoch, e -> e.startOffset));
     }
 
+    @Test
+    public void testReadPasses() throws RemoteStorageException, IOException {
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: the test name should signify what specific scenario are we testing. In this case we are testing that if we don't find any batch in the input stream, then we should return an empty result.

Perhaps rename this to "testReadForMissingFirstBatchInRemote"

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 6]
```diff
@@ -1361,6 +1371,188 @@ private Map<Integer, Long> truncateAndGetLeaderEpochs(List<EpochEntry> entries,
...
+                                                                                    int epochForOffset, long offset) {
+                return Optional.of(segmentMetadata);
+            }
+
+            protected int lookupPositionForOffset(RemoteLogSegmentMetadata remoteLogSegmentMetadata, long offset) {
+                return 1;
+            }
+
+            protected RecordBatch findFirstBatch(RemoteLogInputStream remoteLogInputStream, long offset) throws IOException {
+                return null;
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java:1796]: please add a comment here that this is the key scenario that we are testing here.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 7]
```diff
@@ -1361,6 +1371,188 @@ private Map<Integer, Long> truncateAndGetLeaderEpochs(List<EpochEntry> entries,
...
+        }) {
+            FetchDataInfo fetchDataInfo = remoteLogManager.read(fetchInfo);
+            assertEquals(0, fetchDataInfo.fetchOffsetMetadata.messageOffset);
+            assertEquals(MemoryRecords.EMPTY, fetchDataInfo.records);
+            assertTrue(fetchDataInfo.abortedTransactions.get().isEmpty());
+        }
+    }
+
+    @Test
+    public void testReadPassesWithEmptyRecordsWhenMinOneMessageIsFalse() throws RemoteStorageException, IOException {
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: I would probably rephrase this as:

`testReadForFirstBatchMoreThanMaxFetchBytes`

and then parameterize it using `@ParameterizedTest` with minOneMessage and one without.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 8]
```diff
@@ -1361,6 +1371,188 @@ private Map<Integer, Long> truncateAndGetLeaderEpochs(List<EpochEntry> entries,
...
+
+        FetchRequest.PartitionData partitionData = new FetchRequest.PartitionData(
+                Uuid.randomUuid(), 0, 0, 0, Optional.empty()
+        );
+
+        RemoteStorageFetchInfo fetchInfo = new RemoteStorageFetchInfo(
+                0, false, tp, partitionData, FetchIsolation.HIGH_WATERMARK, false
+        );
+
+        int recordBatchSizeInBytes = 10;
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: please mention here that that this is greater than `fetchMaxBytes`. Even better, create a variable for `fetchMaxBytes` and set this var as `fetchMaxBytes` + 1 which will make it explicit that this is greater.

**Comment by nikramakrishnan** [comment]
Thanks a lot for the review and the suggestions, @divijvaidya. I have addressed your comments and pushed the latest changes to this PR.

**Review by divijvaidya — APPROVED** [review]

**Comment by divijvaidya** [comment]
Thank you for submitting this change @nikramakrishnan. Looks good to me. The test failures in CI are unrelated.
